### PR TITLE
Add in the start of supporting infrastructure for BUFR to IODA direct conversion

### PR DESCRIPTION
**Pull request by CoryMartin-NOAA**
Thanks to all the excellent and hard work by folks like @emilyhcliu , @rmclaren, @nicholasesposito , etc. we are making great progress towards ditching the GSI as our observation preprocessor and converting straight from BUFR to IODA netCDF files.

This PR (will start in draft) modifies the original code I had in the repository which was YAML and `bufr2ioda.x` focused, to start to include the ability to process converter python scripts that use JSON as input.

Note that while there is a templated sample JSON in this PR, there is not yet a sample python script. @emilyhcliu is still working on preparing these files for inclusion in GDASApp.  ---- added end-to-end components for satwind AMVS , completed.  Please see the summary [**here**](https://github.com/NOAA-EMC/GDASApp/pull/575#issuecomment-1678223225).

### CODE REVIEW

**Comment by CoryMartin-NOAA** [comment]
@EdwardSafford-NOAA here's an example of how templated JSON (not YAML) can then be used to create the final configuration files.

**Comment by emilyhcliu** [comment]
I am adding the end-to-end components for GOES AMV in this PR

**Comment by CoryMartin-NOAA** [comment]
Tagging @TravisElless-NOAA for awareness

**Comment by emilyhcliu** [comment]
### Changes made for testing UFO using obs/geoval files from GSI diagnostic files
- **Add option (-q) in GDASApp/ush/ufoeval/run_ufo_hofx_test.sh to run observer without QC filters (only run HofX and obs error assignment)**
- **[Update GDASApp/parm/obs/testing/satwind.yaml](https://github.com/NOAA-EMC/GDASApp/pull/575/files)**
   - Point `GeoVaLs/land_type_index_NPOESS` to `GeoVaLs/water_area_fraction` and `GeoVaLs/land_area_fraction`
   - Modify `ObsFunction/SatWindsSPDBCheck` to `ObsFunction/WindsSPDBCheck` (just drop `sat` from the function name)

- **Update test data set (obs/geoval files) based on GSI latest develop + CRTM-2.4.0_emc**
   - Orion: /work2/noaa.da.eliu/UFO_eval/data/gsi_geovals_l127/nofgat_aug2021/20230811
   - Hera: /scratch1/NCEPDEV/da/Emily.Liu/UFO_eval/data/gsi_geovals_l127/nofgat_aug2021/20230811
   
### Changes made for GOES AMVs end-to-end test
- **Adding [satwind_goes-16.yaml](https://github.com/NOAA-EMC/GDASApp/blob/develop/parm/atm/obs/config/satwind_goes-16.yaml) and [satwind_goes-17.yaml](https://github.com/NOAA-EMC/GDASApp/blob/develop/parm/atm/obs/config/satwind_goes-17.yaml) in GDASApp/parm/obs/config**
   - Filter steps are extracted from ../testing/satwind.yaml
   - Duplicate Check is turned off since it expects GSI usage parameter as input from the ObsSpace --- need to find a way to implement the GSI usage parameter in UFO or find a way to get around it.
   - **Update tropopause pressure estimation**  
      -  Satellite AMV has QC based on tropopause pressure  
      - This check tosses more data than GSI does ([**GOES-17 example with bug**](https://github.com/NOAA-EMC/GDASApp/pull/575#issuecomment-1686429419))
      - [**Bugs** ](https://github.com/NOAA-EMC/GDASApp/pull/575#issuecomment-1684357136)found in the tropopause pressure estimation in fv3-jedi 
      - A PR which fixs the bugs is in PR -[ **fv3-jedi PR #1001**](https://github.com/JCSDA-internal/fv3-jedi/pull/1001) --- under review
      - The data filtering result with the correct tropopause pressure is much closer to GSI ([**GOES-17 example with bug fixed**](https://github.com/NOAA-EMC/GDASApp/pull/575#issuecomment-1686435400))
- **Changes need to be made in the UFO repository to make the end-to-end work properly**
   - UFO filter -  **[ObsErrorFactorPressureCheck](https://github.com/JCSDA-internal/ufo/blob/develop/src/ufo/filters/obsfunctions/ObsErrorFactorPressureCheck.cc)** needs to be generalized to work properly in the end-to-end process
      - Change the required obs error parameter input `adjusted_error_name` to an optional one.   GMAO recently changed to input GSI obs error as a mandatory variable from ObsSpace.  This broke our test.  So. we need to set this parameter to an optional input.  
      - Add an optional variable `geovar_sfc_geomz` for the user to input the desired variable for surface height.  The standard convention in fv3-jedi for surface height is `surface altitude`.  However, some test data sets use `surface_geometric_height` as surface height.  Therefore, need to add the user optional input for flexibility.  
       - These changes are under review in [**UFO PR #2973**](https://github.com/JCSDA-internal/ufo/pull/2973) ---> merged to UFO develop on August 16, 2023

- **Add bufr convert Python driver and json configuration file to GDASApp for GOES AMVs processing**
   - Python driver: GDASApp/ush/ioda/bufr2ioda/[**bufr2ioda_satwind_amv_goes.py**](ush/ioda/bufr2ioda/gen_bufr2ioda_json.py)
   - json configuration: GDASApp/parm/ioda/bufr2ioda/[**bufr2ioda_satwind_amv_goes.json**](parm/ioda/bufr2ioda/bufr2ioda_satwind_amv_goes.json)
   - The Python driver will do the following work:
      - read in json confirmation to set parameters (dump directory, data type, data subset, ...etc)
      - query subset data from BUFR dump file
      - create derived variables
         - wind components from wind direction and speed
         - ObsType group: 240, 241, 245, .....etc (algorithm based)
         - separate GOES data based on platform (satid) --- generate ioda file for goes-16 and goes-17
         - store data in IODA data layout and format
         
### End-to-End test results for GOES-17 AMV:

- **[OmF and Obs Count](https://github.com/NOAA-EMC/GDASApp/pull/575#issuecomment-1686435400)**
- **[Final ObsError](https://github.com/NOAA-EMC/GDASApp/pull/575#issuecomment-1689085428)**





   The end-to-end test of GOES AMVs is documented in [**T2O Issue #72**](https://github.com/NOAA-EMC/JEDI-T2O/issues/72)

File: ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py [hunk 1]
```diff
@@ -0,0 +1,530 @@
+# (C) Copyright 2023 NOAA/NWS/NCEP/EMC
```
Comment (CoryMartin-NOAA) [inline ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py]: Remove the copyright

File: ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+
+import sys
+
+sys.path.append('/work/noaa/da/eliu/JEDI-iodaconv/ioda-bundle/build/lib/')
```
Comment (CoryMartin-NOAA) [inline ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py]: remove hard coded paths for orion, my shell script should handle this properly (I hope)

File: ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+global float32_fill_value
+global int32_fill_value
+global int64_fill_value
+
+float32_fill_value = np.float32(0)
+int32_fill_value = np.int32(0)
+int64_fill_value = np.int64(0)
+
+
+def Compute_WindComponents_from_WindDirection_and_WindSpeed(wdir, wspd):
```
Comment (CoryMartin-NOAA) [inline ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py]: should we put this in a separate py script? I assume many of these converters will use the same function?

File: ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+    satellite_info_array = config["satellite_info"]
+    sensor_name = config["sensor_info"]["sensor_name"]
+    sensor_full_name = config["sensor_info"]["sensor_full_name"]
+    sensor_id = config["sensor_info"]["sensor_id"]
+
+    print('sensor_name = ', sensor_name)
+    print('sensor_full_name = ', sensor_full_name)
+    print('sensor_id = ', sensor_id)
+
+    bufrfile = cycle_type + '.t' + hh + 'z.'+data_type + '.tm00.' + data_format
```
Comment (CoryMartin-NOAA) [inline ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py]: can we use "f-strings" on these lines?

File: ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+    DATA_PATH = dump_dir + '/' + cycle_type + '.' + yyyymmdd + '/' + hh + '/' + bufrfile
+
+    # ============================================
+    # Make the QuerySet for all the data we want
+    # ============================================
+    start_time = time.time()
+
+    print('Making QuerySet ...')
+#   q = bufr.QuerySet()
```
Comment (CoryMartin-NOAA) [inline ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py]: remove this commented out line

File: ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+    q.add('hour', '*/HOUR')
+    q.add('minute', '*/MINU')
+    q.add('second', '*/SECO')
+    q.add('sensorZenithAngle', '*/SAZA')
+    q.add('sensorCentralFrequency', '*/SCCF')
+    q.add('pressure', '*/PRLC[1]')
+
+    # Processing Center
+    q.add('dataProviderOrigin', '*/OGCE[1]')
+#   q.add('windGeneratingApplication', '*/AMVQIC/GNAPS')
```
Comment (CoryMartin-NOAA) [inline ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py]: do we need the commented out lines in the next few sections?

File: ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+    wspd = r.get('windSpeed')
+
+    print(' ... Executing QuerySet: get datatime: observation time ...')
+    # DateTime: seconds since Epoch time
+    # IODA has no support for numpy datetime arrays dtype=datetime64[s]
+    timestamp = r.get_datetime('year', 'month', 'day', 'hour', 'minute', 'second').astype(np.int64)
+
+    print(' ... Executing QuerySet: Done!')
+
+    print(' ... Executing QuerySet: Check BUFR variable generic dimension and type ...')
```
Comment (CoryMartin-NOAA) [inline ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py]: We should try to use the `wxflow` logger and make some of this stuff "debug" and some "info"

File: ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
+    print('     pressure  shape = ', pressure.shape)
+    print('     wdir      shape = ', wdir.shape)
+    print('     wspd      shape = ', wspd.shape)
+
+    print('     timestamp type  = ', timestamp.dtype)
+    print('     satid     type  = ', satid.dtype)
+    print('     lon       type  = ', lon.dtype)
+    print('     chanfreq  type  = ', chanfreq.dtype)
+    print('     swcm      type  = ', swcm.dtype)
+    print('     satzenang type  = ', satzenang.dtype)
```
Comment (CoryMartin-NOAA) [inline ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py]: Do we need all of these prints or is this just for prototype development?

File: ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py [hunk 9]
```diff
@@ -0,0 +1,530 @@
...
+            print('     cvwd2      shape = ', cvwd2.shape)
+
+            print(' ... Create ObsSpae for', satinst, 'satid = ', sat)
+            # Create the dimensions
+            dims = {
+                'Location': np.arange(0, wdir2.shape[0])
+            }
+
+            # Create IODA ObsSpace
+            iodafile = cycle_type + '.t' + hh + 'z.' + data_type + '.' + satinst + '.tm00.nc'
```
Comment (CoryMartin-NOAA) [inline ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py]: Would like to use f-strings again on these lines

File: ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py [hunk 10]
```diff
@@ -0,0 +1,530 @@
...
+
+
+if __name__ == '__main__':
+
+    start_time = time.time()
+
+    with open("bufr2ioda_ncep_satwind_amv_goes.json", "r") as json_file:
+        config = json.load(json_file)
+
+    print('emily checking config = ', config)
```
Comment (CoryMartin-NOAA) [inline ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py]: need this one removed/changed to just say "config ="

File: ush/ufoeval/run_ufo_hofx_test.sh [hunk 11]
```diff
@@ -65,7 +70,13 @@ obtype=$1
 machine=${machine:-orion} 
 
 if [ $machine = orion ]; then
-   workdir=/work2/noaa/da/$LOGNAME/ufoeval/$cycle/$obtype
+#  workdir=/work2/noaa/da/$LOGNAME/ufoeval/$cycle/$obtype
```
Comment (CoryMartin-NOAA) [inline ush/ufoeval/run_ufo_hofx_test.sh]: can remove instead of comment out

File: ush/ufoeval/run_ufo_hofx_test.sh [hunk 12]
```diff
@@ -65,7 +70,13 @@ obtype=$1
...
+   if [ $run_filtering == NO ]; then
+      workdir=/work2/noaa/da/$LOGNAME/ufoeval/$cycle/${obtype}_noqc
+      echo "Run without data filtering"
+   else
+      workdir=/work2/noaa/da/$LOGNAME/ufoeval/$cycle/${obtype}
+   fi
    GDASApp=${GDASApp:-/work2/noaa/da/$LOGNAME/git/GDASApp/} # Change this to your own branch
 elif [ $machine = hera ]; then
```
Comment (CoryMartin-NOAA) [inline ush/ufoeval/run_ufo_hofx_test.sh:80]: should support filtering on hera too

**Review by CoryMartin-NOAA — COMMENTED** [review]
@emilyhcliu I have a few comments about the python script, basically just cleanup and some printing/logging comments. Let's discuss after the 2pm meeting, I'm happy to take the lead on most/all of the changes I propose.

File: ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py (hunk 3 above)
Comment (CoryMartin-NOAA) [reply ush/ioda/bufr2ioda/bufr2ioda_ncep_satwind_amv_goes.py]: we can tackle this in a future PR

**Comment by CoryMartin-NOAA** [comment]
If someone would like to test this out themselves (at least the BUFR to IODA part of it):
`module use /path/to/gdasapp/modulefiles`
`module load GDAS/${machine}`
`cd ush/ioda/bufr2ioda`
`./run_bufr2ioda.sh $PDY$cyc $DUMP /scratch1/NCEPDEV/global/glopara/dump $GDASApp/parm/ioda/bufr2ioda $STMP/output_ioda`

**Comment by CoryMartin-NOAA** [comment]
Also note, the plan would be for `run_bufr2ioda.sh` to be called as part of the prep job in the global-workflow, reading in BUFR from the dumps and putting IODA files in the COMROT

**Comment by emilyhcliu** [comment]
### Issue - Tropopause PressureCheck for Satellite Wind AMVs
- This check rejects obs when pressure is more than 50 mb above the tropopause.
- In the end-to-end test, the pressure check rejects more data than GSI 
- This indicates that tropopause estimation is different between GSI and UFO.  The one in UFO has bugs

GSI|UFO
--- | ---
gsi_tropopause_pressure_time1|ufo_tropopause_pressure_time1 (4)

A fix for tropopause calculation is proposed to[ **fv3-jedi repository PR #1001**](https://github.com/JCSDA-internal/fv3-jedi/pull/1001)

**Here is the updated tropopause pressure from UFO**
GSI|UFO
--- | ---
gsi_tropopause_pressure_time1|ufo_tropopause_pressure_time1 (3)

**Comment by emilyhcliu** [comment]
### Approximately 1500+ data are screened out due to incorrect tropopause pressure from fv3-jedi:

ufo_abi_goes-17_omf_windEastward_qc_time1
gsi_abi_goes-17_omf_windEastward_qc_time1

**Comment by emilyhcliu** [comment]
### The result of filtering wth the correct tropopause pressure estimation from fv3-jedi is much closer to GSI

ufo_abi_goes-17_omf_windEastward_qc_time1 (1)
gsi_abi_goes-17_omf_windEastward_qc_time1 (1)

**Review by RussTreadon-NOAA — COMMENTED** [review]
Changes seem reasonable, but I did not dig into the details. Does it make sense to add hera or orion GW labels to run these changes through CI testing? Perhaps not. My initial attempt to build `feature/bufr2ioda_example` on Hera failed with a soca error

```
-- Adding bundle project soca
-- Cloning soca from https://github.com/jcsda-internal/soca.git into /scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/soca...
-- /scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/soca retrieved.
-- Updating soca to head of BRANCH feature/change_mask_value...
-- git fetch --all @ /scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/soca
-- git fetch --all --tags @ /scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/soca
-- git checkout feature/change_mask_value @ /scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/soca
CMake Error at /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.1/envs/unified-env/install/intel/2021.5.0/ecbuild-3.7.2-mf7qlax/share/ecbuild/cmake/ecbuild_log.cmake:190 (message):
  ^[[1;35mCRITICAL - git checkout feature/change_mask_value on
  /scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/soca
  failed:

    /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.1/envs/unified-env/install/intel/2021.5.0/git-2.38.1-37lxdk2/bin/git checkout -q feature/change_mask_value
    error: pathspec 'feature/change_mask_value' did not match any file(s) known to git

  ^[[m
Call Stack (most recent call first):
  /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.1/envs/unified-env/install/intel/2021.5.0/ecbuild-3.7.2-mf7qlax/sha\
re/ecbuild/cmake/ecbuild_git.cmake:221 (ecbuild_critical)
  /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.1/envs/unified-env/install/intel/2021.5.0/ecbuild-3.7.2-mf7qlax/sha\
re/ecbuild/cmake/ecbuild_bundle.cmake:157 (ecbuild_git)
  CMakeLists.txt:97 (ecbuild_bundle)


-- Configuring incomplete, errors occurred!
See also "/scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/build/CMakeFiles/CMakeOutput.log".
See also "/scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/build/CMakeFiles/CMakeError.log".
```
Anyone else see this error?

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA yes that was fixed in #576 sorry!

**Comment by RussTreadon-NOAA** [comment]
Pulled [15ef730](https://github.com/NOAA-EMC/GDASApp/commit/15ef7306f5fd082c4dd75a6c0fa5ab20cae317e2) into my working copy and will rerun the build.   Thank you, @CoryMartin-NOAA , for the quick action.

**Comment by RussTreadon-NOAA** [comment]
`feature/bufr2ioda_example` at [15ef730](https://github.com/NOAA-EMC/GDASApp/commit/15ef7306f5fd082c4dd75a6c0fa5ab20cae317e2) successfully built on Hera.

Execution of 
```
test
./run_bufr2ioda.sh 2023082100 gdas /scratch1/NCEPDEV/global/glopara/dump /scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/parm/ioda/bufr2ioda /scratch2/NCEPDEV/stmp1/role.jedipara/output_ioda
```
failed with error

```
FileNotFoundError: [Errno 2] No such file or directory: '/scratch2/NCEPDEV/stmp1/Russ.Treadon/output_ioda/satwind_amv_goes_2023082100.json'
Problem running converter script for satwind_amv_goes
```

Realtime `DMPDIR` paths include `atmos`.

Make the following change to `gdas.cd/ush/ioda/bufr2ioda/bufr2ioda_satwind_amv_goes.py`
```
-    DATA_PATH = os.path.join(dump_dir, f"{cycle_type}.{yyyymmdd}", str(hh), bufrfile)
+    DATA_PATH = os.path.join(dump_dir, f"{cycle_type}.{yyyymmdd}", str(hh), 'atmos', bufrfile)
```
Rerun command with following result

```
(gdasapp) Hera(hfe04):/scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/ush/ioda/bufr2ioda$ ./run_bufr2ioda.sh 2023082100 gdas /scratch1/NCEPDEV/global/glopara/dump /scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/parm/ioda/bufr2ioda /scratch2/NCEPDEV/stmp1/role.jedipara/output_ioda
Processing satwind_amv_goes...
2023-08-21 18:56:57,329 - INFO     - gen_bufr2ioda_json.py: Using /scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/parm/ioda/bufr2ioda/bufr2ioda_satwind_amv_goes.json as input
2023-08-21 18:56:57,341 - INFO     - gen_bufr2ioda_json.py: Wrote to /scratch2/NCEPDEV/stmp1/role.jedipara/output_ioda/satwind_amv_goes_2023082100.json
2023-08-21 18:56:58,022 - INFO     - BUFR2IODA_satwind_amv_goes.py: input config =
2023-08-21 18:56:58,022 - INFO     - BUFR2IODA_satwind_amv_goes.py: sensor_name = ABI
2023-08-21 18:56:58,022 - INFO     - BUFR2IODA_satwind_amv_goes.py: sensor_full_name = Advanced Baseline Imager
2023-08-21 18:56:58,023 - INFO     - BUFR2IODA_satwind_amv_goes.py: sensor_id = 617
2023-08-21 18:56:58,023 - INFO     - BUFR2IODA_satwind_amv_goes.py: Making QuerySet
2023-08-21 18:56:58,024 - INFO     - BUFR2IODA_satwind_amv_goes.py: Executing QuerySet to get ResultSet
2023-08-21 18:59:29,510 - INFO     - BUFR2IODA_satwind_amv_goes.py: Executing QuerySet: get metadata
2023-08-21 19:00:38,946 - INFO     - BUFR2IODA_satwind_amv_goes.py: Executing QuerySet Done!
2023-08-21 19:00:38,946 - INFO     - BUFR2IODA_satwind_amv_goes.py: Running time for executing QuerySet to get ResultSet : 220.92170119285583 seconds
2023-08-21 19:00:38,946 - INFO     - BUFR2IODA_satwind_amv_goes.py: Creating derived variables - wind components (uob and vob)
2023-08-21 19:00:39,397 - INFO     - BUFR2IODA_satwind_amv_goes.py: Running time for creating derived variables : 0.45093464851379395 seconds
2023-08-21 19:00:39,397 - INFO     - BUFR2IODA_satwind_amv_goes.py: Split data based on satellite id, Create IODA ObsSpace and Write IODA output
2023-08-21 19:00:39,549 - INFO     - BUFR2IODA_satwind_amv_goes.py: Number of Unique satellite identifiers: 2
2023-08-21 19:00:39,549 - INFO     - BUFR2IODA_satwind_amv_goes.py: Unique satellite identifiers: [270 272]
2023-08-21 19:00:39,661 - INFO     - BUFR2IODA_satwind_amv_goes.py: Number of Unique observation timestamp: 6
2023-08-21 19:00:39,661 - INFO     - BUFR2IODA_satwind_amv_goes.py: Create ObsSpae for abi_goes-16 satid = 270
2023-08-21 19:00:39,679 - INFO     - BUFR2IODA_satwind_amv_goes.py: Create output file: /scratch2/NCEPDEV/stmp1/role.jedipara/output_ioda/gdas.t00z.satwnd.abi_goes-16.tm00.nc
2023-08-21 19:00:42,412 - INFO     - BUFR2IODA_satwind_amv_goes.py: Running time for splitting and output IODA for abi_goes-16: 2.862285852432251 seconds
2023-08-21 19:00:42,573 - INFO     - BUFR2IODA_satwind_amv_goes.py: Number of Unique observation timestamp: 6
2023-08-21 19:00:42,573 - INFO     - BUFR2IODA_satwind_amv_goes.py: Create ObsSpae for abi_goes-18 satid = 272
2023-08-21 19:00:42,575 - INFO     - BUFR2IODA_satwind_amv_goes.py: Create output file: /scratch2/NCEPDEV/stmp1/role.jedipara/output_ioda/gdas.t00z.satwnd.abi_goes-18.tm00.nc
2023-08-21 19:00:48,094 - INFO     - BUFR2IODA_satwind_amv_goes.py: Running time for splitting and output IODA for abi_goes-18: 5.681908845901489 seconds
2023-08-21 19:00:48,094 - INFO     - BUFR2IODA_satwind_amv_goes.py: All Done!
2023-08-21 19:00:56,880 - INFO     - BUFR2IODA_satwind_amv_goes.py: Total running time: 238.85879802703857 seconds
(gdasapp) Hera(hfe04):/scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/ush/ioda/bufr2ioda$
```
Directory `/scratch2/NCEPDEV/stmp1/role.jedipara/output_ioda` contains the following files
```
-rw-r--r-- 1 role.jedipara stmp 30539253 Aug 21 19:00 gdas.t00z.satwnd.abi_goes-16.tm00.nc
-rw-r--r-- 1 role.jedipara stmp 60833956 Aug 21 19:00 gdas.t00z.satwnd.abi_goes-18.tm00.nc
```

Next step is to pass files in `output_ioda` to fv3-jedi var, letkf, or hofx app.

**Comment by CoryMartin-NOAA** [comment]
Thank you @RussTreadon-NOAA for testing, I will add `atmos` and push the commit.

I can confirm that FV3-JEDI HofX works on GOES-16 AMVs using the YAML provided here. Approximately 1.9 million obs (965k each u,v) processed in ~3 mins on 6 nodes, 1 PE each at C768 background resolution.
Can look at: `/scratch1/NCEPDEV/stmp2/Cory.R.Martin/tmp_aug2023/work_hofx_test` if you'd like

**Review by RussTreadon-NOAA — DISMISSED** [review]
Changes  look good to me.   Successfully ran `run_bufr2ioda.sh` for 2023082100 gdas.  @CoryMartin-NOAA confirmed fv3-jedi hofx works with output.

Approve.

**Review by RussTreadon-NOAA — APPROVED** [review]
Refresh my working copy of `feature/bufr2ioda_example` on Hera.  Execute
```
(gdasapp) Hera(hfe09):/scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/ush/ioda/bufr2ioda$ ./run_bufr2ioda.sh 2023082112 gdas /scratch1/NCEPDEV/global/glopara/dump /scratch1/NCEPDEV/da/role.jedipara/git/global-workflow/test1/sorc/gdas.cd/parm/ioda/bufr2ioda scratch2/NCEPDEV/stmp1/role.jedipara/output_ioda
```

Script ran to completion after about 2 minutes of execution.   Create two output files; `gdas.t12z.satwnd.abi_goes-16.tm00.nc` and `gdas.t12z.satwnd.abi_goes-18.tm00.nc`.

Will we, at some point, replace python processing with c++ processing.  I assume c++ will be faster.   Alternatively, might fv3-jedi app be able to directly read bufr files via a c++ interface?

Approve changes.

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA C++ might be faster, the underlying code here is compiled. While it's a Python wrapper, the BUFR library is Fortran, and the IODA API, C++.  My hope is that eventually, IODA will read BUFR directly, but that might still be quite some time. For now, I think 2-3 mins, with using something like CMP, should be fast enough for prototyping.

**Comment by emilyhcliu** [comment]
> @RussTreadon-NOAA C++ might be faster, the underlying code here is compiled. While it's a Python wrapper, the BUFR library is Fortran, and the IODA API, C++. My hope is that eventually, IODA will read BUFR directly, but that might still be quite some time. For now, I think 2-3 mins, with using something like CMP, should be fast enough for prototyping.

@RussTreadon-NOAA @CoryMartin-NOAA Cory gave a good summary about the BUFR converter. @rmclaren and I will meet this afternoon to discuss the backend implementation for BUFR data.  My understanding is that the BUFR backend can be applied to the Python wrapper with some additional work in the IODA.

**Comment by emilyhcliu** [comment]
**Just want to add the `Final ObsError` plots (UFO vs. GSI) from the end-to-end test  --- Looks close enough**

ufo_abi_goes-17_obserr_windEastward_qc_time1

gsi_abi_goes-17_obserr_windEastward_qc_time1

**Pull request merged**